changefeedccl: support CREATE DATABASE CHANGEFEED syntax

KeithCh opened the pull request:
Add basic syntax support for database-level 
changefeeds.
`CREATE DATABASE CHANGEFEED for foo`
This PR only allows the aforementioned statement 
to be parsed; no changefeed is created.

Epic: [CRDB-1421](https://cockroachlabs.atlassian.net/browse/CRDB-1421)
Resolves: https://github.com/cockroachdb/cockroach/issues/147369

Release note: None

KeithCh commented on pkg/sql/parser/sql.y:
```diff
@@ -6215,18 +6217,40 @@ create_stats_option:
+// FOR <target> [INTO sink] [WITH <options>]
+//
+// sink: data capture stream destination (Enterprise only)
+create_database_changefeed_stmt:
```
Created a separate non-terminal token for the sole purpose of having proper help text.
e.g. when running `CREATE DATABASE CHANGEFEED ??`. If we didn't have this it would always show the `CREATE DATABASE` help text.

There is a case where there is a conflict with the `CREATE DATABASE` help text, where it'll output that instead.
e.g. `CREATE DATABASE for CHANGEFEED ??`

andyyang890 commented on pkg/sql/sem/tree/changefeed.go:
```diff
@@ -10,9 +10,16 @@ import (
+type ChangefeedLevel string
+
+const (
+	ChangefeedLevelDatabase ChangefeedLevel = "DATABASE"
```
Instead of using strings, maybe this should be an enum where the zero value is `ChangefeedLevelTable`

andyyang890 commented on pkg/sql/sem/tree/changefeed.go:
```diff
@@ -27,14 +34,17 @@ func (node *CreateChangefeed) Format(ctx *FmtCtx) {
 	} else {
 		// Sinkless feeds don't really CREATE anything, so the syntax omits the
 		// prefix. They're also still EXPERIMENTAL, so they get marked as such.
 		ctx.WriteString("EXPERIMENTAL ")
```
Should we maybe take this opportunity to get rid of the `EXPERIMENTAL` for sinkless feeds?

KeithCh replied:
Filed an issue here cause we didn't have one before #147663.

I'll leave it out of this PR since it's big enough already.

andyyang890 commented on pkg/sql/parser/sql.y:
```diff
@@ -547,11 +547,11 @@ func (u *sqlSymUnion) grantTargetList() tree.GrantTargetList {
     return u.val.(*tree.GrantTargetList)
 }
-func (u *sqlSymUnion) changefeedTargets() tree.ChangefeedTargets {
+func (u *sqlSymUnion) changefeedTableTargets() tree.ChangefeedTargets {
```
this should still be named `changefeedTargets` right

KeithCh replied:
I renamed it because it's guaranteed that every target in the list will be table targets, but I can rename it to match the type.

KeithCh replied:
Actually I'm in favor of keeping it just so it is easy to know what type of targets are being used.

andyyang890 commented on pkg/sql/parser/sql.y:
```diff
@@ -1254,6 +1254,7 @@ func (u *sqlSymUnion) doBlockOption() tree.DoBlockOption {
 %type <tree.Statement> create_stmt
 %type <tree.Statement> create_schedule_stmt
 %type <tree.Statement> create_changefeed_stmt create_schedule_for_changefeed_stmt
+%type <tree.Statement> create_database_changefeed_stmt create_schedule_for_changefeed_stmt
```
I don't think you should be duplicating `create_schedule_for_changefeed_stmt` from the line above
```suggestion
%type <tree.Statement> create_database_changefeed_stmt
```

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -528,7 +532,11 @@ func createChangefeedJobRecord(
 	tableOnlyTargetList := tree.BackupTargetList{}
 	for _, t := range changefeedStmt.Targets {
-		tableOnlyTargetList.Tables.TablePatterns = append(tableOnlyTargetList.Tables.TablePatterns, t.TableName)
+		target, ok := t.(*tree.ChangefeedTableTarget)
```
i don't love having to do these assertions everywhere. is this (having target be an interface) the right abstraction?

KeithCh replied:
I guess if we don't plan on supporting multiple types of targets per changefeed it's not really needed.

KeithCh replied:
One benefit is that we don't have to modify the signature of functions that takes in tree.ChangefeedTargets, and the way we loop through them. If we didn't have the abstraction we would have to do something like 
```
type ChangefeedTargets struct {
 ChangefeedDatabseTarget ChangefeedDatabseTarget[]
 ChangefeedTableTarget ChangefeedTableTargets[]
}
```

KeithCh replied:
That is possible, but we can't rely on FamilyName to determine whether it is a DB target. We might have `with split_column_families` set in which case it would be empty. It would have to be like

```
type ChangefeedTarget struct {
	// Name is the name of a database or table.
	Name
	// FamilyName is the name of a column family to be watched. It is the empty
	// string for database-level targets.
	FamilyName Name
	
	// TargetType is an enum (table, changefeed)
	TargetType ChangefeedTargetType
}
```
and then we check the `TargetType` when we work with the object.

But that doesn't seem very idomatic.

KeithCh replied:
Talked to @asg0451, we're going to have multiple fields inside `CreateChangefeed` to represent the different target types.

asg0451 reviewed (commented):
some thoughts; also can you get someone from sql foundations to actually review the parser changes?

mgartner commented on pkg/sql/parser/sql.y:
```diff
@@ -6215,6 +6217,28 @@ create_stats_option:
+//
+// sink: data capture stream destination (Enterprise only)
+create_database_changefeed_stmt:
+  CREATE DATABASE CHANGEFEED FOR database_name opt_changefeed_sink opt_with_options
```
Did you attempt lexer changes that would allow us to do `CREATE CHANGEFEED FOR DATABASE`? What issues did you run into? I worry that we're going to regret confusing customers with `CREATE DATABASE CHANGEFEED` and `CREATE CHANGEFEED FOR TABLE` and it breaks the pattern for nearly every other database object where you create a `<thing>` with `CREATE <thing> ...`. I think a little extra complexity on our side is worth it in the long-run.

KeithCh replied:
Used your code from https://github.com/cockroachdb/cockroach/pull/148116 🙏

mgartner commented on pkg/sql/sem/tree/changefeed.go:
```diff
@@ -10,9 +10,17 @@ import (
+
+const (
+	ChangefeedLevelTable    ChangefeedLevel = iota
+	ChangefeedLevelDatabase ChangefeedLevel = iota + 1
```
nit: You can omit the type and `= iota + 1` for all entries after the first, e.g.,: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/sem/tree/with.go#L26-L31

nit: briefly describe the type and each value

msbutler commented on pkg/ccl/changefeedccl/changefeedbase/target.go:
```diff
@@ -14,7 +14,7 @@ import (
 // Target provides a version-agnostic wrapper around jobspb.ChangefeedTargetSpecification.
 type Target struct {
 	Type              jobspb.ChangefeedTargetSpecification_TargetType
-	TableID           descpb.ID
```
some motivation: if i git blame LDR code, it's weird to see  a change to `changefeedccl: support CREATE DATABASE CHANGEFEED syntax`.

KeithCh replied:
that makes sense

KeithCh committed:
changefeedccl: support CREATE DATABASE CHANGEFEED syntax

Add basic syntax support for database-level
changefeeds.
CREATE DATABASE CHANGEFEED for foo
This PR only allows the aforementioned statement
to be parsed; no changefeed is created.

Epic: CRDB-1421
Resolves: #147369

Release note: None

KeithCh commented on pkg/cmd/docgen/diagrams.go:
```diff
@@ -784,7 +784,7 @@ var specs = []stmtSpec{
 	{
 		name:   "create_changefeed_stmt",
-		inline: []string{"changefeed_targets", "opt_changefeed_sink", "opt_with_options", "kv_option_list", "kv_option"},
+		inline: []string{"changefeed_table_targets", "opt_changefeed_sink", "opt_with_options", "kv_option_list", "kv_option"},
```
https://github.com/cockroachdb/cockroach/issues/149347

asg0451 commented on pkg/sql/parser/sql.y:
```diff
@@ -6194,12 +6195,22 @@ create_stats_option:
 // sink: data capture stream destination (Enterprise only)
 create_changefeed_stmt:
-  CREATE CHANGEFEED FOR changefeed_targets opt_changefeed_sink opt_with_options
+  CREATE CHANGEFEED FOR changefeed_table_targets opt_changefeed_sink opt_with_options 
```
` `

asg0451 commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -394,27 +394,27 @@ func generateAndValidateNewTargets(
 	sinkURI string,
 ) (
-	tree.ChangefeedTargets,
+	tree.ChangefeedTableTargets,
```
can you add an if level = db then unimplemented err pls

KeithCh replied:
I'm going to put this in the next PR since I don't have the changes here that will tell us that a changefeed being altered is a database-level changefeed. That information will live in ChangefeedTargetSpecification, and I haven't added the different target types.
